frontend/src/editor/speaker_dropdown.tsx

```diff
@@ -1,83 +1,78 @@
-import { useMemo, useState } from 'react';
-import { useSlate, ReactEditor } from 'slate-react';
+import { memo, useState, useCallback } from 'react';
+import { Editor } from 'slate';
+import { useSlateStatic, ReactEditor } from 'slate-react';
 import * as Automerge from '@automerge/automerge';
-import { AutomergeEditor } from 'slate-automerge-doc';
 
 import { Document, Paragraph } from './types';
 import { PrimaryButton, SecondaryButton } from '../components/button';
 import { IoIosCreate, IoIosList, IoIosTrash } from 'react-icons/io';
 import { Dropdown, DropdownItem, DropdownSection } from '../components/dropdown';
 import { Input, Select } from '../components/form';
+import { useSpeakerName, useSpeakerNames } from '../utils/document';
 import { showModal, Modal, ModalProps } from '../components/modal';
 
-export function getSpeakerName(element: Paragraph, speaker_names: Record<string, string>): string {
-  if (!element.speaker) {
-    return `Unknown Speaker`;
-  } else if (element.speaker in speaker_names) {
-    return speaker_names[element.speaker];
-  } else {
-    return `Unnamed Speaker ${element.speaker}`;
-  }
-}
-
 function SelectSpeakerModal({
-  doc,
+  editor,
   onSpeakerSelected,
   onNewSpeaker,
   onClose,
   selected,
   ...props
 }: {
-  doc: Automerge.Doc<Document>;
+  editor: Editor;
   onSpeakerSelected: (speakerId: string) => void;
   onNewSpeaker: (speakerName: string) => void;
   onClose: () => void;
-  selected?: string;
+  selected: string | null;
 } & Omit<ModalProps, 'label'>) {
   const NEW_SPEAKER_OPTION = '__new_speaker';
 
   const [speakerId, setSpeakerId] = useState<string | null>(null);
   const [speakerName, setSpeakerName] = useState<string>('');
 
-  const speakerNames = useMemo(() => {
-    const spkNames: Record<string, string> = {};
-    for (const para of doc.children) {
-      if (para.speaker !== null && !(para.speaker in spkNames)) {
-        spkNames[para.speaker] = getSpeakerName(para, doc.speaker_names);
-      }
-    }
-    return spkNames;
-  }, [doc.speaker_names]);
+  const speakerNames = useSpeakerNames(editor);
 
-  if (speakerId === null && selected !== undefined && selected in speakerNames) {
+  if (speakerId === null && selected !== null && selected in speakerNames) {
     setSpeakerId(selected);
   } else if (speakerId === null && Object.keys(speakerNames).length > 0) {
     setSpeakerId(Object.keys(speakerNames)[0]);
   } else if (speakerId === null && Object.keys(speakerNames).length === 0) {
     setSpeakerId(NEW_SPEAKER_OPTION);
   }
```

Comment:
This breaks the "change speaker" modal: Since with `useSpeakerNames` `speakerNames` is `{}` initially and only filled with a value after a short delay, this will always go to the last branch of this if condition